Kafka/Confluent with Airflow - Basic tutorial

TJaniF opened the pull request:
Drafting what I learn about Kafka/Confluent with some caveats:

-The installation of the provider differs between M1 and others, both my devices are M1 so I'd need someone to test if I got the instructions for non-M1 right.
-I'm testing the code with a cluster in the Confluent cloud. I don't think anything should be different for connection to local Kafka but haven't tested it yet because I ran into a fun little problem with local Kafka - Java does not want to install so far 😅

kentdanas commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,279 @@
+        xcom_push_key="retrieved_message",
+    )
+
+    consumer_task >> await_message
```
Shouldn't the `producer_task` created in Step 2 still be in this DAG?

kentdanas reviewed (commented):
@TJaniF I think this will be a helpful tutorial! This is a request the field gets a lot, so I think having simple instructions like this will be very useful. 

Left a couple of initial comments here. I haven't gotten the DAG working with my local kafka yet, but we can look at that next week. And I do agree with you that having a # How it works section at the end with more operator information would be helpful.

TJaniF reviewed (commented):
Thank you for the review! I added some more information on the operator and their parameters. :)

dylanbstorey commented:
@TJaniF 

I'd probably put a brief discussion section/admonition at the beginning of the article to help warn people off of attempting to use this provider for any sort of streaming or low latency work.

TJaniF commented:
Changed the install instructions, they now work with 6.0.3, 6.0.4 and 7.0.0-alpha2 for me (on an M1 Mac but that shouldnt make a difference for not-standalone instances). 

I also added a note at the start warning to follow best practices, as well as a best practices section. :)

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+
+The [Airflow Kafka provider package](https://github.com/astronomer/airflow-provider-kafka) contains an operator allowing you to produce events directly to an existing Kafka topic, the ProduceToTopicOperator.
+
+1. Create a new Python file in your `dags` folder called `kafka_example_dag_1.py`.
```
```suggestion
1. Create a new file in your `dags` folder called `kafka_example_dag_1.py`.
```

jwitz commented on learn/airflow-kafka.md:
````diff
@@ -0,0 +1,345 @@
+        )
+    ```
+
+    The code above retrieves the environment variables you defined in [Step 1](#step-1-configure-your-astro-project) and packages them into a configuration dictionary that can be used by the ProduceToTopicOperator. Any Python function which returns a generator can be passed to the `producer_function` parameter of the ProduceToTopicOperator. Make sure your producer function returns a generator tha ...
````
```suggestion
    The code above retrieves the environment variables you defined in [Step 1](#step-1-configure-your-astro-project) and packages them into a configuration dictionary that can be used by the ProduceToTopicOperator. Any Python function which returns a generator can be passed to the `producer_function` parameter of the ProduceToTopicOperator. Make sure your producer function returns a generator that contains key-value pairs where the value is in a format your Kafka topic accepts as input. In this example, the generator produces a JSON value. Additionally, if you have defined a schema for your Kafka topic, the generator needs to return compatible objects.
```

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+
+    The code above retrieves the environment variables you defined in [Step 1](#step-1-configure-your-astro-project) and packages them into a configuration dictionary that can be used by the ProduceToTopicOperator. Any Python function which returns a generator can be passed to the `producer_function` parameter of the ProduceToTopicOperator. Make sure your producer function returns a generator tha ...
+
+    If you are connecting to a local Kafka cluster you might need to adjust the `connection_config` dictionary (see tip below).
```
```suggestion
```

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+
+:::tip
+
+If you are using a Kafka cluster that is running locally on your machine you will need to adjust the `connection_config` parameter. To connect to the [Kafka quick start](https://kafka.apache.org/documentation/#quickstart) cluster when running Airflow in Docker, set the following properties in your Kafka cluster's `server.properties` file:
```
```suggestion
If you are using a Kafka cluster that is running on your local machine, you will need to adjust the `connection_config` parameter in this example. To connect to the [Kafka quick start](https://kafka.apache.org/documentation/#quickstart) cluster when running Airflow in Docker, set the following properties in your Kafka cluster's `server.properties` file:
```

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+
+If you are using a Kafka cluster that is running locally on your machine you will need to adjust the `connection_config` parameter. To connect to the [Kafka quick start](https://kafka.apache.org/documentation/#quickstart) cluster when running Airflow in Docker, set the following properties in your Kafka cluster's `server.properties` file:
+
+listeners=PLAINTEXT://:9092,DOCKER_HACK://:19092
+advertised.listeners=PLAINTEXT://localhost:9092,DOCKER_HACK://host.docker.internal:19092
+listener.security.protocol.map=PLAINTEXT:PLAINTEXT,DOCKER_HACK:PLAINTEXT
```
Add ` ```text ` styling here

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+advertised.listeners=PLAINTEXT://localhost:9092,DOCKER_HACK://host.docker.internal:19092
+listener.security.protocol.map=PLAINTEXT:PLAINTEXT,DOCKER_HACK:PLAINTEXT
+
+while using `"bootstrap.servers":"host.docker.internal:19092"` and `"security.protocol": "PLAINTEXT"` as your `connection_config`. Learn more in [Confluent's documentation](https://www.confluent.io/blog/kafka-client-cannot-connect-to-broker-on-aws-on-docker-etc/#scenario-5).
```
```suggestion
Then add `"bootstrap.servers":"host.docker.internal:19092"` and `"security.protocol": "PLAINTEXT"` to your `connection_config` parameter in the DAG.
```

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+
+## Step 3: Add a consumer task
+
+With the ConsumeFromTopicOperator Airflow is able to consume messages from topics. 
```
```suggestion
The ConsumeFromTopicOperator enables Airflow to consume messages from topics. 
```

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+
+With the ConsumeFromTopicOperator Airflow is able to consume messages from topics. 
+
+1. Add the following import statement to the `kafka_example_dag_1`.
```
```suggestion
1. Add the following import statement to `kafka_example_dag_1`.
```

jwitz commented on learn/airflow-kafka.md:
````diff
@@ -0,0 +1,345 @@
+    from airflow_provider_kafka.operators.consume_from_topic import ConsumeFromTopicOperator
+    ```
+
+2. In the `kafka_example_dag_1` add the following code below the `producer_task`.
````
```suggestion
2. Add the following code after the `producer_task` in `kafka_example_dag_1`.
```

jwitz commented on learn/airflow-kafka.md:
````diff
@@ -0,0 +1,345 @@
+
+    ```
+
+    In this code snippet first a function is defined that reads from the topic defined as `my_topic` and prints the messages consumed to the Airflow task log. The ConsumeFromTopicOperator uses the same `connection_config` as the `producer_task` with added configurations specific to Kafka Consumers. You can read more about consumer configuration in Kafka in the official [Kafka documentation](https ...
````
```suggestion
   The `consumer_task` includes a function that reads from `my_topic` and prints the messages it consumes to the Airflow task log. The `consumer_task` uses the same `connection_config` as the `producer_task` with added configurations specific to Kafka Consumers. You can read more about consumer configuration in Kafka in the [Kafka documentation](https://kafka.apache.org/documentation/#consumerconfigs).
```

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+
+    ![Consumer log](/img/guides/kafka-consumer-logs.png)
+
+:::info
```
```suggestion
:::tip
```

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+
+    ![Producer logs](/img/guides/confluent-produced-tasks.png)
+
+:::tip
```
```suggestion
:::info
```

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+
+:::info
+
+A very common pattern is to directly connect an Amazon S3 bucket to your Kafka topic as a consumer. The ConsumeFromTopicOperator is helpful if you want to use Airflow features to schedule the consuming task. Instead of writing the messages retrieved to the Airflow logs, you can write them to S3 using the [S3CreateObjectOperator](https://registry.astronomer.io/providers/amazon/modules/s3createobje ...
```
```suggestion
A common use case is to directly connect an Amazon S3 bucket to your Kafka topic as a consumer. The ConsumeFromTopicOperator is helpful if you want to use Airflow to schedule the consuming task. Instead of writing the messages retrieved to the Airflow logs, you can write them to S3 using the [S3CreateObjectOperator](https://registry.astronomer.io/providers/amazon/modules/s3createobjectoperator).
```

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+
+A common use case is to run a downstream task when a specific message appears in your Kafka topic. The AwaitKafkaMessageOperator is a deferrable operator that will listen to your Kafka topic for a message that fulfills a specific criteria.
+
+:::info
+
+A deferrable operator is a sensor that will go into a deferred state in between checking for its condition in the target system. While in the deferred state the operator does not take up a worker slot, offering a significant efficiency improvement. Using a deferrable operator necessitates the presence of the Triggerer component. Learn more about [Deferrable operators]((https://docs.astronomer.io/ ...
+
+:::
```
```suggestion
A deferrable operator is a sensor that will go into a deferred state in between checking for its condition in the target system. While in the deferred state the operator does not take up a worker slot, offering a significant efficiency improvement. See [Deferrable operators]((https://docs.astronomer.io/learn/deferrable-operators)).
```

jwitz commented on learn/airflow-kafka.md:
````diff
@@ -0,0 +1,345 @@
+    consumer_task >> await_message
+    ```
+
+    In this code snippet an `await_function` is defined which will parse each message in the Kafka topic and return the message if the value is an integer divisible by 5. The AwaitKafkaMessageOperator using runs this function over messages polled from the Kafka topic. If no matching message is found it goes into a deferred state until it tries again.  
````
```suggestion
    This code snippet includes an `await_function` which will parse each message in the Kafka topic and return the message if the value is an integer divisible by 5. The AwaitKafkaMessageOperator using runs this function over messages polled from the Kafka topic. If no matching message is found, it continues to poll in a deferred state until it finds one.
```

jwitz replied:
Is the last sentence right? I thought that it's always trying even if it's in a deferred state.

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+- `KafkaConsumerHook`
+- `KafkaProducerHook`
+
+as well as three operators and one trigger:
```
```suggestion
It uses three operators and one trigger:
```

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+- `AwaitKafkaMessageOperator`
+- `AwaitMessageTrigger`
+
+in this section we will provide more detailled information on the parameters of the operators. For more information on the other modules in this provider see the [`airflow-provider-kafka`](https://github.com/astronomer/airflow-provider-kafka) source code.
```
```suggestion
The following section provides more detailed information on the parameters of each operator. For more information on the other modules in this provider see the [`airflow-provider-kafka`](https://github.com/astronomer/airflow-provider-kafka) source code.
```

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+
+The ProduceToTopicOperator can be used create a Kafka producer to produce messages to a Kafka topic. You can define the following parameters:
+
+- `topic`: the Kafka topic you want to produce to.
```
Capitalize the first letter of all these descriptions.

jwitz reviewed (approved):
Looks good to me! Just minor wording suggestions throughout.

jwitz commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,345 @@
+id: airflow-kafka
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
```
```suggestion
```

kentdanas commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,333 @@
+
+:::info
+
+If you are using a Kafka cluster that is running on your local machine, you will need to adjust the `connection_config` parameter in this example. To connect to the [Kafka quick start](https://kafka.apache.org/documentation/#quickstart) cluster when running Airflow in Docker, set the following properties in your Kafka cluster's `server.properties` file:
```
I found this part a little confusing since it's instructing to update the parameter directly in the `connection_config` in the DAG, but in the DAG code we're pulling the value from an env variable that we create in Step 1. Wondering if this note should be a step above, maybe as part of Step 1. Or below substep 2 in this step and be super clear that we're saying make the adjustment directly to the DAG file.

TJaniF replied:
Yeah you are right, I changed the config so that all parts that need adjustments are passed as env variables and put the note in Step 1, as well as the note about how to change server.properties in the Prerequisites, I feel this is a bit more logical, so that the prereq is "Kafka server running with correct configs" before step 1. I hope it is clearer now :)

kentdanas commented on learn/airflow-kafka.md:
````diff
@@ -0,0 +1,333 @@
+
+If you are using a Kafka cluster that is running on your local machine, you will need to adjust the `connection_config` parameter in this example. To connect to the [Kafka quick start](https://kafka.apache.org/documentation/#quickstart) cluster when running Airflow in Docker, set the following properties in your Kafka cluster's `server.properties` file:
+
+```text
````
Not sure if this would be included in the Kafka knowledge prereq (which I definitely don't have), but I got tripped up for a minute because I added these lines to `server.properties` after my cluster was already running, and didn't realize I needed to restart the cluster for them to take effect. I'll leave it up to you whether we should add something like "before starting your cluster" to the end of the last sentence, or whether you think any Kafka users would already know this :)

TJaniF replied:
Very good point. I had forgotten to restart too at first 😅

kentdanas commented on learn/airflow-kafka.md:
```diff
@@ -0,0 +1,333 @@
+
+## Best practices
+
+Apache Kafka is a tool optimized for streaming messages at high frequencies for example in an IoT application. Airflow is designed to handle orchestration of data pipelines in batches at a frequency of once per minute or less.
```
```suggestion
Apache Kafka is a tool optimized for streaming messages at high frequencies, for example in an IoT application. Airflow is designed to handle orchestration of data pipelines in batches at a frequency of no more than once per minute.
```

kentdanas reviewed (approved):
Looks great! A few very minor comments, but otherwise I think this one is good to go!

TJaniF reviewed (commented):
Thank you @kentdanas :).

TJaniF commented:
Waiting to merge because Viraj and I want to run the blog post plus this tutorial together by Kafka experts before publishing. :)
